ext/dom/html_document.c

```
@@ -0,0 +1,1047 @@
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "p|ls!", &filename, &filename_len, &options, &override_encoding, &override_encoding_len) == FAILURE) {
+		RETURN_THROWS();
+	}
+
+	/* See php_libxml_streams_IO_open_wrapper(), apparently this caused issues in the past. */
+	if (strstr(filename, "%00")) {
+		zend_argument_value_error(1, "must not contain percent-encoded NUL bytes");
+		RETURN_THROWS();
+	}
+
+	if (!check_options_validity(options)) {
+		RETURN_THROWS();
+	}
+
+	dom_lexbor_libxml2_bridge_application_data application_data;
+	application_data.input_name = filename;
+	application_data.current_total_offset = 0;
+	application_data.html_no_implied = options & HTML_PARSE_NOIMPLIED;
+	dom_reset_line_column_cache(&application_data.cache_tokenizer);
+	lexbor_libxml2_bridge_parse_context ctx;
+	lexbor_libxml2_bridge_parse_context_init(&ctx);
+	if (!(options & XML_PARSE_NOERROR)) {
+		lexbor_libxml2_bridge_parse_set_error_callbacks(&ctx, dom_lexbor_libxml2_bridge_tokenizer_error_reporter, dom_lexbor_libxml2_bridge_tree_error_reporter);
+	}
+	ctx.application_data = &application_data;
+
+	char buf[4096];
+
+	/* Setup everything encoding & decoding related */
+	dom_decoding_encoding_ctx decoding_encoding_ctx;
+	dom_decoding_encoding_ctx_init(&decoding_encoding_ctx);
+	bool should_determine_encoding_implicitly = true; /* First read => determine encoding implicitly */
+	if (override_encoding != NULL) {
+		const lxb_encoding_data_t *encoding_data = lxb_encoding_data_by_name((const lxb_char_t *) override_encoding, override_encoding_len);
+		if (!encoding_data) {
+			zend_argument_value_error(3, "must be a valid document encoding");
+			RETURN_THROWS();
+		}
+		should_determine_encoding_implicitly = false;
+		dom_setup_parser_encoding_manually((const lxb_char_t *) buf, encoding_data, &decoding_encoding_ctx, &application_data);
+	}
+
+	stream = php_stream_open_wrapper_ex(filename, "rb", REPORT_ERRORS, /* opened_path */ NULL, /* context */ php_libxml_get_stream_context());
```

Won't this emit warnings?

You might need to replace the error handler here, something I do hate but IDK how else to deal with stream errors.

---

I only really found ext/spl doing that for streams, other exts don't seem to do that? Not sure what's best?

---

Wait... no ignore me, the reason SPL does it, is because constructors must either construct a valid object or throw an exception. But here it's just a static method.

---

Well, I do throw an exception if none has been thrown later on in this method, if none has been thrown by the stream wrapper: lines 720-722. This is an approach I've seen elsewhere too, just spl is indeed the _exception_ ;p

---

Right, I might try, as a future thing, to make the stream layer to throw exceptions instead of warnings, but this will be tricky :(